### PR TITLE
SDL: Add GameController API, cleanup

### PR DESCRIPTION
**Pull request by SuperSamus**
Dolphin uses the Joystick API on the SDL backend. However, this API doesn't have consistent mapping across controllers.
For instance, here is the same mapping between a Nintendo Switch Pro Controller and an Xbox controller:
image
image

This is extremely cumbersome when having to deal with multiple controllers, since it means you need to have a profile per controller. (And it makes Dolphin significantly less plug and play: for example, if a friend comes over with a new controller, here comes the need to remap everything for them.)

The GameController API instead has consistent mapping across all controllers it supports:
image

I also made sure that the names are the same from XInput, to allow for more profile portability (by default it uses `Button N/E/S/W`, but it supports `Button Y/B/A/X`).

Now, the notes:
- ~This breaks every old SDL mapping (obviously).~
  - Now it uses GameController when possible and Joystick for compatibility with old configurations and weird controllers.
- I deleted a lot of workarounds indiscriminately. I don't know whether they are still needed or not.
  - Also, CMake now forces SDL to be at least version 2.26. This means that all the `ifdefs` are removed.
  - The PR needs more testing in general, I only tested on Linux.
- Also, SDL input will now be enabled by default on Linux. Every other emulator does this (including Dolphin itself on Flatpak), and it makes motion controls much more plug and play (I don't think anyone wants to follow [this guide](https://wiki.dolphin-emu.org/index.php?title=Motion_evdev) when a friend comes over with a new controller).
  - However, it seems to make Dolphin's GUI startup much slower (by at least a second). Maybe related to #12335? 

Closes: https://bugs.dolphin-emu.org/issues/13164

### CODE REVIEW

**Comment by mbc07** [comment]
@dolphin-emu-bot rebuild

**Comment by mbc07** [comment]
Gave it a try on Windows, with a DualShock 3, a DualShock 4 and a generic XInput controller, while this seem (overall) an improvement, there are regressions:

* Dolphin is now very prone to randomly crash when setting inputs (especially when opening the Advanced Input Editor)
* SDL, as setup in this PR, doesn't allow individually selecting the rumble motors anymore, you get only one "Motor" binding that controls all rumble motor the controller have

Also, I don't think forcing the "Button A/B/X/Y" names for the face buttons (always following Xbox layout) for all controllers is a good idea, especially when the three major players in this space have their own layouts (e.g. reversed A/B and X/Y on Nintendo, and simply Square, Cross, Triangle and Circle on Sony). I'd rather keep using generic names for them (like the current "Button 1/2/3/4" from master), just making sure they always point to the same face buttons on every controller, to allow reusing the same profile across different controllers...

**Comment by SuperSamus** [comment]
> Dolphin is now very prone to randomly crash when setting inputs (especially when opening the Advanced Input Editor)

Can't reproduce on Linux.
I'll semi-revert the SDL version bump (from 2.28.2 to 2.26.5) because it doesn't build on old Mac versions, _maybe_ it'll fix the crashes on your end?
Otherwise, I'm probably going to need logs.

> SDL doesn't allow individually selecting the rumble motors anymore

I admit I trimmed too much stuff... I'll put it back if someone else agrees that it needs to be put back.

> Also, I don't think forcing the "Button A/B/X/Y" names for the face buttons (always following Xbox layout) for all controllers is a good idea

My motivation is that, by being a superset of XInput, you can easily migrate from using XInput to SDL without changing a thing.
Most profiles I see online are made using XInput (e.g. [Super Mario Galaxy](https://forums.dolphin-emu.org/Thread-super-mario-galaxy-1-or-2-xinput-controller-profile), [Metroid Other M](https://forums.dolphin-emu.org/Thread-metroid-other-m-xbox360-controller-config)).

> to allow reusing the same profile across different controllers

You can do that, `Button A` for instance is simply the south button, whether it's an Xbox, Nintendo or Playstation controller.
Still, I'll wait for more feedback.

**Comment by mbc07** [comment]
@dolphin-emu-bot rebuild

**Comment by mbc07** [comment]
Tried the most recent version of this PR again and it's still prone to randomly crash (but a bit less than before, I think?). I've ran the debug version but didn't find anything too meaningful in the log, got this when opening Dolphin:

```
49:56:599 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: Incorrect parameter.
49:56:649 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: Unexpected controller element crc
49:56:687 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:687 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:687 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:687 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:689 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: Unexpected controller element crc
49:56:689 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:689 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:689 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:689 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:689 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:689 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:690 InputCommon\ControllerInterface\DualShockUDPClient\DualShockUDPClient.cpp:392 I[CI]: DualShockUDPClient Restart
49:56:702 InputCommon\ControllerInterface\ControllerInterface.cpp:302 N[CI]: Added device: DInput/0/Keyboard Mouse
49:56:732 InputCommon\ControllerInterface\ControllerInterface.cpp:302 N[CI]: Added device: DInput/0/Wireless Controller
49:56:732 InputCommon\ControllerInterface\ControllerInterface.cpp:302 N[CI]: Added device: XInput/0/Gamepad
49:56:733 InputCommon\ControllerInterface\ControllerInterface.cpp:302 N[CI]: Added device: WGInput/0/Controlador de jogo compatível com HID
49:56:733 InputCommon\ControllerInterface\ControllerInterface.cpp:302 N[CI]: Added device: WGInput/0/Controlador XBOX 360 para Windows
49:56:733 InputCommon\ControllerInterface\DualShockUDPClient\DualShockUDPClient.cpp:487 I[CI]: DualShockUDPClient PopulateDevices
49:56:734 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: Unexpected controller element crc
49:56:747 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:747 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:747 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:747 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:747 InputCommon\ControllerInterface\ControllerInterface.cpp:302 N[CI]: Added device: SDL/0/PS4 Controller
49:56:747 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: Unexpected controller element crc
49:56:747 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:747 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:747 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:747 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:747 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:747 InputCommon\ControllerInterface\SDL\SDL.cpp:144 D[CI]: error: That operation is not supported
49:56:747 InputCommon\ControllerInterface\ControllerInterface.cpp:302 N[CI]: Added device: SDL/0/Xbox 360 Controller
```

After that, the log is just an endless spam of `InputCommon\ControllerInterface\WGInput\WGInput.cpp:528 D[CI]: WGInput: UpdateBatteryLevel failed.` until Dolphin eventually crashes.

-----

> I admit I trimmed too much stuff... I'll put it back if someone else agrees that it needs to be put back.

While the other points are open to discussion (especially since this PR still is a draft), I'd say that's pretty much a blocker for this getting merged. You're actually removing existing functionality (mapping individual rumbles) for no good reason.

**Comment by SuperSamus** [comment]
SDL3 is considering to use N S W E for the new API, so that's an argument in favor for not using A B X Y on Dolphin.
Still, I do want the easy migration from XInput... maybe there could be a simple alias system?

Anyway, I'm waiting for more input (pun intended).

**Comment by cobalt2727** [comment]
Any way around the huge SDL2 version bump? Ubuntu 20.04, for example, is still on [2.0.10](https://packages.ubuntu.com/search?suite=focal&searchon=names&keywords=libsdl2)

**Comment by jordan-woyak** [comment]
Users who want to stay with an older version of their operating system should have no issues staying with an older version of the emulator.

**Comment by SuperSamus** [comment]
SDL is bundled in Dolphin, so that shouldn't be a problem (and there's even a PR to update it too, #12144).

**Comment by SuperSamus** [comment]
Alright, I restored everything that existed before, and I think that this can now be reviewed for real.

I tested it for a bit, but I can't test devices that only support Joystick, nor devices where some buttons are supported by GameController but others are only supported by Joystick. (They should work, though)

**Comment by mbc07** [comment]
@dolphin-emu-bot rebuild

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by SuperSamus** [comment]
With the last force push, now it checks that the controller is a GameController before calling `SDL_GameControllerName`.
If it's not, it calls `SDL_JoystickName` instead.
Now, not-GameControllers should actually work.

**Comment by Filoppi** [comment]
Code LGTM, if it was up to me, I'd merge :). Haven't tried this though, maybe somebody could confirm these problems have been fixed: https://github.com/dolphin-emu/dolphin/pull/12085#issuecomment-1665135602

File: Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp
Comment (jordan-woyak) [inline Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp:257]: I don't think calling `SDL_CONTROLLER_BUTTON_A` `B` `X` `Y`, "S" "E" "W" "N" is a good idea. Can we really assume those are the physical positions of those buttons? Don't even Nintendo gamepads differ here? What does SDL do for Sony gamepads?

File: Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp
Comment (jordan-woyak) [inline Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp:398]: Instead of adding "not detectable" legacy inputs. I think it would be cleaner for the newly introduced GameController `Input`s to override `IsMatchingName` to match the old "Button 1" naming. This would still allow loading old configs but without flooding the input list with redundant entries. This is how input backend naming changes have been handled in the past.

Comment (SuperSamus) [reply Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp:398]: I tried that, but doing that is a mess.
Just because something is a Button in the GameController API doesn't mean it's a Button in the Joystick API too.
Axes have different ranges (-32768 to 32767 normally, but are from 0 to 32767 *only* in the GameController API).

It's not worth it.

All the legacy inputs are added at the end, so that they are at the bottom of the list.

**Comment by SuperSamus** [comment]
> Code LGTM, if it was up to me, I'd merge :). Haven't tried this though, maybe somebody could confirm these problems have been fixed: [#12085 (comment)](https://github.com/dolphin-emu/dolphin/pull/12085#issuecomment-1665135602)

They are, I restored everything.
(I can't test if it crashes on Windows, so that's the only point remaining)

File: Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp
Comment (jordan-woyak) [reply Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp:257]: Are you saying a switch pro controller's physical button B will activate "Button S" despite being called `SDL_CONTROLLER_BUTTON_A`?

Comment (SuperSamus) [reply Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp:257]: Exactly.

**Comment by SuperSamus** [comment]
Anything else that I need to do before this can be merged?
(Note: Windows still needs to be tested)

**Review by mbc07 — CHANGES_REQUESTED** [review]
Just tested the most recent version of this PR on Windows, and the following issues still persists:
* Dolphin stills randomly freeze, then crash shortly after, when binding inputs (happens more frequently when opening the Advanced Input Editor or after clicking "Refresh" in the device list)
* SDL, as set up in this PR, still doesn't allow individually selecting the rumble motors anymore, you get only one "Motor" binding that simultaneously activate all rumble motors the controller have

**Comment by SuperSamus** [comment]
> Dolphin stills randomly freeze, then crash shortly after, when binding inputs (happens more frequently when opening the Advanced Input Editor or after clicking "Refresh" in the device list)

What if you restore the Windows-related `#ifdef`s?

> SDL, as set up in this PR, still doesn't allow individually selecting the rumble motors anymore, you get only one "Motor" binding that simultaneously activate all rumble motors the controller have

They appear and work on my end.
image

**Comment by mbc07** [comment]
> What if you restore the Windows-related `#ifdef`s?

This might sound a bit harsh, but I don't think I should be patching other people's PRs in hope they don't break currently working behavior from the master branch for no good reason. If you think this might solve the issue, update your PR with those changes baked in and I'll gladly review it again.
<br/>

> They appear and work on my end. 
> image

What you're seeing are rumble effects, not individual motors. On most modern gamepads, you should have at least separate "Motor L" and "Motor R" bindings (on top of the effects shown in your screenshot) to reach the same functionality of the master branch, which allows independently controlling the available rumble motors of the controller.

**Comment by SuperSamus** [comment]
> update your PR with those changes baked in and I'll gladly review it again.

Done. For now it's a separate commit, I'll squash it later if it does the trick.
 
> What you're seeing are rumble effects, not individual motors. On most modern gamepads, you should have at least separate "Motor L" and "Motor R" bindings (on top of the effects shown in your screenshot) to reach the same functionality of the master branch, which allows independently controlling the available rumble motors of the controller.

Then it was like this even before, it's not a regression.

**Comment by mbc07** [comment]
> Done. For now it's a separate commit, I'll squash it later if it does the trick.

Well, it seems to have worked. Messed around for about 10 minutes, not a single crash.
<br/>

> Then it was like this even before, it's not a regression.

Yes, it is. I just checked Dolphin 5.0-20814 and I have "Strong" and "Weak" listed in the motor selection when accessing the gamepad via SDL (not "Motor L" and "Motor R" like I mentioned before, but they work the same, allowing to control the motors separatedly). With the current version of this PR, I get only "Motor" in the list, which will enable both at the same time.

-----
TL;DR, random crashes on Windows seems fixed, rumble regression still there...

**Comment by SuperSamus** [comment]
> I have "Strong" and "Weak" listed in the motor selection

Weird, I restored everything that should make them appear...

Still because newer controller already didn't support anything about SDL Haptics, I added new `Motor L` and `Motor R` that use the GameController API, and do... what the name says.
`Motor` still remains: on SDL, every rumble command cancels the previous one.

> random crashes on Windows seems fixed

Thanks for the testing!

**Comment by mbc07** [comment]
@dolphin-emu-bot rebuild

**Review by mbc07 — APPROVED** [review]

**Comment by SuperSamus** [comment]
One last question: there was (and still is) some inconsistency in the `GetName()` method: some classes take the string from their `m_name`, others compute the string on that spot.
Which one should I do?
1. Who cares, leave it as it is
2. Every class has an `m_name`
3. No class has an `m_name`, all strings are computed in `GetName()`

**Comment by jordan-woyak** [comment]
> One last question: there was (and still is) some inconsistency in the `GetName()` method: some classes take the string from their `m_name`, others compute the string on that spot. Which one should I do?

I don't think it matters. The function is only called for configuration loading/saving, not regularly. Do whatever is appropriate for the particular `Input`.

I do still think effort should be made to merge the current and legacy inputs, though. :(

**Comment by SuperSamus** [comment]
> I do still think effort should be made to merge the current and legacy inputs, though.

As I said before, doing so with `IsMatchingName` would be a giant mess.

Maybe an alternative would be a `IsHidden` method that hides the input from the UI?

That still isn't super straightforward. (e.g. does a true `IsHidden` imply a false `IsDetectable`? In that case, how is the behavior maintened if the latter is overridden?)

Any opinions? (Including whether any of this should be done at all)

(btw, I won't be able to work on this for a week)

File: Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp
Comment (jordan-woyak) [inline Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp:393]: It seems like `AddAnalogInputs` isn't being used correctly here, which isn't difficult because the naming could be better.

`AddAnalogInputs` is specifically for handling inputs that have an unknown neutral position. An example is DirectInput or "legacy" SDL axes which might represent an analog stick with a neutral position of 0 or potentially a trigger with a neutral position of e.g. -32768. `AddAnalogInputs` adds additional "Full Surface" inputs to handle the case of the trigger that uses the full axis value range.

If we know the type of input and the neutral position for certain, which seems to be the case here with the "GameController" interface, just `AddInput` can be used. Adding the "Full Surface" inputs for a analog stick is unnecessary. It's only needed when something might be a trigger that uses the full negative-to-positive range.

**Comment by infinitymdm** [comment]
Just tested this patch on linux with a Switch Pro Controller. Motion works out-of-the-box, just like compiling with -DENABLE_SDL. The options for evdev input methods still appear in the controller select dropdown too.

Is there anything in particular that needs to be tested? I'll be glad to help out however I can to get this merged.

**Comment by SuperSamus** [comment]
> Is there anything in particular that needs to be tested?

AFAIK, gamepads that aren't supported by the GameController API and are only supported by the old Joystick API that has always been here. It's not impossible that I may have missed an `if (gamecontroller)` somewhere.
macOS also hasn't been tested.

Still, the blocker for this PR is [this thread](https://github.com/dolphin-emu/dolphin/pull/12085#discussion_r1425728793). I'd want to discuss an alternative way to solve this (because implementing all of this using `IsMatchingName` was how I tried doing it first, but it is _such_ a mess), but no one is saying anything.

**Comment by AdmiralCurtiss** [comment]
Poking @jordan-woyak, I don't think anyone else is super familiar with the Input stuff.

Maybe it would be enough to just hide the old button names from the input list via some `hidden` flag or something? Or do you have something more sophisticated in mind?

**Comment by AdmiralCurtiss** [comment]
Also: Is this something that needs to be solved in this PR or can it be done separately? This has been sitting here for a while in a practically finished state, from what I understand.

**Comment by SuperSamus** [comment]
> Maybe it would be enough to just hide the old button names from the input list via some hidden flag or something?

[It is fine by me](https://github.com/dolphin-emu/dolphin/pull/12085#issuecomment-1867481726), but I wanted more feedback before moving forward with that.

> Is this something that needs to be solved in this PR[...]?

I don't really consider these extra inputs to be an issue, so for me the answer is no.
But if many others consider it a problem, then it's a yes. So... does anyone else consider it a problem?

**Comment by infinitymdm** [comment]
IMO the ease of setting up motion controls with SDL vastly outweighs extra clutter in input lists. If it's a problem, it can be fixed in another PR so as not to impede these improvements.

**Comment by jordan-woyak** [comment]
I think it's fine to resolve the input list clutter in another PR.

**Comment by AdmiralCurtiss** [comment]
I gave this a test before hitting the button and noticed that the Y axis on the sticks is inverted compared to our other backends, ie. `Y Axis+` is Up on XInput/WGInput but Down in the new SDL. Should we invert that or just accept whatever SDL gives us?

**Comment by SuperSamus** [comment]
Because I added this:

```
bool GameController::Button::IsMatchingName(std::string_view name) const
{
  if (GetName() == name)
    return true;

  // So that SDL can be a superset of XInput
  if (name == "Button A")
    return GetName() == "Button S";
  if (name == "Button B")
    return GetName() == "Button E";
  if (name == "Button X")
    return GetName() == "Button W";
  if (name == "Button Y")
    return GetName() == "Button N";
  return false;
}
```

I guess I should also invert the axes on SDL side.

File: Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp
Comment (AdmiralCurtiss) [inline Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp:578]: Whatever this evaluates as (even I'm not sure...), VS emits a warning for it due to unclear operator precedence.

You may just want to take the more obvious approach and do something like

```
  if (m_axis >= 4)
  {
    // The triggers are only positive, and must not have a sign
  return std::string(s_sdl_axis_names[m_axis]);
  }

  bool negative = m_range < 0;

  // Respect XInput: the vertical axes (where m_axis % 2 == 1) are inverted on SDL
  if (m_axis % 2 == 1)
    negative = !negative;

  return std::string(s_sdl_axis_names[m_axis]) + (negative ? '-' : '+');
```

**Pull request merged**
